### PR TITLE
Update `mmdeviceapi`

### PR DESCRIPTION
**Pull request by trlim**
This PR is to add full binding of `mmdeviceapi` and its dependent module `um/propkeydef`.

### CODE REVIEW

**Comment by retep998** [comment]
PR has conflicts.

**Comment by retep998** [comment]
PR has conflicts.

**Comment by trlim** [comment]
Fixed.

**Comment by retep998** [comment]
PR has conflicts again.

**Comment by retep998** [comment]
Thanks github. Because any significant enough commit is going to modify `build.rs` that means merging most PRs will cause a conflict for most other PRs due to their conservative conflict resolution. If I merge without the web API then I can't squash the commits otherwise github doesn't mark the PR as merged. Sigh.

PR has conflicts.

**Comment by trlim** [comment]
Resolved.

File: src/um/mmdeviceapi.rs [hunk 1]
```diff
@@ -63,10 +191,45 @@ interface IMMDeviceEnumerator(IMMDeviceEnumeratorVtbl): IUnknown(IUnknownVtbl) {
...
+RIDL!(#[uuid(0x72a22d78, 0xcde4, 0x431d, 0xb8, 0xcc, 0x84, 0x3a, 0x71, 0x19, 0x9b, 0x6d)]
+interface IActivateAudioInterfaceAsyncOperation(IActivateAudioInterfaceAsyncOperationVtbl): IUnknown(IUnknownVtbl) {
+    fn GetActivateResult( 
+        activateResult: *mut HRESULT,
+        activatedInterface: *mut *mut IUnknown
+    ) -> HRESULT
+}
+);
+
+EXTERN!{stdcall fn ActivateAudioInterfaceAsync(
```
Comment (retep998) [inline src/um/mmdeviceapi.rs]: `EXTERN!` is gone now.

File: src/um/mmdeviceapi.rs [hunk 2]
```diff
@@ -5,17 +5,70 @@
...
+use shared::minwindef::{DWORD, LPARAM, LPVOID, UINT};
+// use shared::winerror::{ERROR_NOT_FOUND, ERROR_UNSUPPORTED_TYPE, HRESULT_FROM_WIN32};
+use shared::wtypes::PROPERTYKEY;
 use um::propidl::PROPVARIANT;
 use um::propsys::IPropertyStore;
 use um::unknwnbase::{IUnknown, IUnknownVtbl};
 use um::winnt::{HRESULT, LPCWSTR, LPWSTR};
+
+#[repr(C)]
+pub struct MMDeviceEnumerator;
```
Comment (retep998) [inline src/um/mmdeviceapi.rs]: Is this forward declaration necessary?

File: src/shared/guiddef.rs [hunk 3]
```diff
@@ -5,8 +5,9 @@
 // All files in the project carrying such notice may not be copied, modified, or distributed
 // except according to those terms.
 //! GUID definition
+// Done as of 10.0.14393.0
 use ctypes::{c_uchar, c_ulong, c_ushort};
-STRUCT!{struct GUID {
+STRUCT!{#[derive(PartialEq, Eq)] struct GUID {
```
Comment (retep998) [inline src/shared/guiddef.rs]: I'd rather not derive any traits at this time.

File: src/macros.rs [hunk 4]
```diff
@@ -46,6 +46,24 @@ macro_rules! DEFINE_GUID {
...
 }
 #[macro_export]
+macro_rules! DEFINE_PROPERTYKEY {
+    (
+        $name:ident, $l:expr, $w1:expr, $w2:expr,
+        $b1:expr, $b2:expr, $b3:expr, $b4:expr, $b5:expr, $b6:expr, $b7:expr, $b8:expr,
+        $pid:expr
+    ) => {
+        pub const $name: PROPERTYKEY = PROPERTYKEY {
+            fmtid: GUID {
```
Comment (retep998) [inline src/macros.rs]: `$crate::shared::guiddef::GUID`

File: build.rs [hunk 5]
```diff
@@ -113,7 +114,9 @@ const DATA: &'static [(&'static str, &'static [&'static str], &'static [&'static
     ("pdh", &["basetsd", "guiddef", "minwindef", "winnt", "wtypesbase", "windef"], &["pdh"]),
     ("processthreadsapi", &["basetsd", "minwindef", "winnt"], &["kernel32"]),
     ("propidl", &["minwindef", "wtypes"], &[]),
+    ("propkeydef", &["guiddef", "wtypes"], &[]),
     ("propsys", &["unknwnbase"], &[]),
+    ("propsys", &["minwindef", "propidl", "propkeydef", "unknwnbase", "winnt", "wtypes"], &[]),
```
Comment (retep998) [inline build.rs]: Why is propsys defined twice?

**Pull request merged**